### PR TITLE
Add `IndexedJsonDocument`, a `JSONWrapper` implementation that stores JSON documents in a prolly tree with probabilistic hashing.

### PR DESCRIPTION
**Pull request by nicktobey**
tl;dr: We store a JSON document in a prolly tree, where the leaf nodes of the tree are blob nodes with each contain a fragment of the document, and the intermediate nodes are address map nodes, where the keys describe a JSONPath.

The new logic for reading and writing JSON documents is cleanly separated into the following files:

IndexedJsonDocument - The new `JSONWrapper` implementation. It holds the root hash of the prolly tree.

JsonChunker - A wrapper around a regular chunker. Used to write new JSON documents or apply edits to existing documents.

JsonCursor - A wrapper around a regular cursor, with added functionality allowing callers to seek to a specific location in the document.

JsonScanner - A custom JSON parser that tracks that current JSONPath.

JsonLocation - A custom representation of a JSON path suitable for use as a prolly tree key.

Each added file has additional documentation with more details about the individual components.

Throughout every iteration of this project, the core idea has always been to represent a JSON document as a mapping from JSONPath locations to the values stored at those locations, then we could store that map in a prolly tree and get all the benefits that we currently get from storing tables in prolly trees: fast diffing and merging, fast point lookups and mutations, etc.

This goal has three major challenges:
- For deeply nested JSON documents, simply listing every JSONPath requires asymptotically more space than the original document.
- We need to do this in a way that doesn't compromise performance on simply reading JSON documents from a table, which I understand is the most common use pattern.
- Ideally, users should not need to migrate their databases, or update their clients in order to read newer dbs, or have to choose between different configurations based on their use case.

This design achieves all three of these requirements:
- While it requires additional storage, this additional storage cannot exceed the size of the original document, and is in practice much smaller.
- It has indistinguishable performance for reading JSON documents from storage, while also allowing asymptotically faster diff and merge operations when the size of the changes is much smaller than the size of the document. (There is a cost: initial inserts of JSON documents are currently around 20% slower, but this is a one-time cost that does not impact subsequent reads and could potentially be optimized further.)
- Documents written by the new `JSONChunker` are backwards compatible with current Dolt binaries and can be read back by existing versions of Dolt. (Although they will have different hashes than equivalent documents that those versions would write.)

### CODE REVIEW

File: go/store/prolly/tree/json_chunker.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	if err != nil {
+		return Node{}, err
+	}
+
+	jsonChunker, err := newEmptyJsonChunker(ctx, ns)
+	if err != nil {
+		return Node{}, err
+	}
+	jsonChunker.appendJsonToBuffer(jsonBytes)
+	jsonChunker.processBuffer(ctx)
```
Comment (reltuk) [inline go/store/prolly/tree/json_chunker.go:57]: Handler `err` return.

File: go/store/prolly/tree/json_chunker.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		}
+		return j.chunker.Done(ctx)
+	}
+	cur := j.jCur.cur
+	cursorDecoder := j.jCur.jsonScanner
+	jsonBytes := cursorDecoder.jsonBuffer[cursorDecoder.previousValueOffset:]
+	// Append the rest of the JsonCursor, then continue until we either exhaust the cursor, or we coincide with a boundary from the original tree.
+	for {
+		j.appendJsonToBuffer(jsonBytes)
+		j.processBuffer(ctx)
```
Comment (reltuk) [inline go/store/prolly/tree/json_chunker.go:167]: Error check...

File: go/store/prolly/tree/json_chunker.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+				return err
+			}
+			var newBuffer []byte
+			if !j.jScanner.atEndOfChunk() {
+				newBuffer = j.jScanner.jsonBuffer[j.jScanner.valueOffset:]
+			}
+			newScanner := ScanJsonFromMiddle(newBuffer, j.jScanner.currentPath)
+			j.jScanner = &newScanner
+		}
+	}
```
Comment (reltuk) [inline go/store/prolly/tree/json_chunker.go:242]: Is it more clear to just keep the start offset explicitly rather than resetting j.jScanner each time?

```go
	startOffset := 0
	for j.jScanner.AdvanceToNextLocation() != io.EOF {
		key := j.jScanner.currentPath.key
		value := j.jScanner.jsonBuffer[startOffset:j.jScanner.valueOffset]
		if crossesBoundary(key, value) {
			err := j.createNewLeafChunk(ctx, key, value)
			if err != nil {
				return err
			}
			startOffset = j.jScanner.valueOffset
		}
	}
	j.jScanner = ScanJsonFromMiddle(j.jScanner.jsonBuffer[startOffset:], j.jScanner.currentPath)
	return nil
```

File: go/store/prolly/tree/json_chunker.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+func (j *JsonChunker) writeKey(keyPath jsonLocation) {
+	finalPathElement, isArray := keyPath.getLastPathElement()
+
+	isFirstValue := j.jScanner.firstElementOrEndOfEmptyValue()
+
+	if !isFirstValue {
+		j.appendJsonWithoutSplitting([]byte{','})
+	}
+	if !isArray {
+		j.appendJsonWithoutSplitting([]byte(fmt.Sprintf(`"%s":`, finalPathElement)))
```
Comment (reltuk) [inline go/store/prolly/tree/json_chunker.go:126]: Definitely somewhat confusing on how this fits together with the existing jScanner state and it's currentPath, and why it's safe specifically in this one case immediately after the chunker has been constructed with this specific keyPath and never safe again.

**Review by reltuk — COMMENTED** [review]
Generally looks awesome and a really cool approach. Not too mind bending after starting at it for a while. I took a pass on some suggestions, but I may be missing some things. Please push back if I'm off base anywhere :).

File: go/store/prolly/tree/node_cursor.go [hunk 5]
```diff
@@ -217,6 +217,9 @@ func newLeafCursorAtKey[K ~[]byte, O Ordering[K]](ctx context.Context, ns NodeSt
 // searchForKey returns a SearchFn for |key|.
 func searchForKey[K ~[]byte, O Ordering[K]](key K, order O) SearchFn {
 	return func(nd Node) (idx int) {
+		if nd.keys.IsEmpty() {
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/node_cursor.go:220]: These are the flattened leaf nodes. They contain 0 keys and 1 value. If we don't have this check, then the binary search will attempt to compare the search key to key[0] in the node, which will be out of bounds.

I added a comment.

File: go/store/prolly/tree/json_location.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		return 1
+	}
+	if firstByte <= 248 {
+		return 2
+	}
+	return int(firstByte - 246)
+}
+
+// jsonPathElementsFromMySQLJsonPath computes a jsonLocation from a MySQL JSON path (https://dev.mysql.com/doc/refman/8.0/en/json.html#json-path-syntax)
+func jsonPathElementsFromMySQLJsonPath(pathBytes []byte) (path jsonLocation, err error) {
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: I like this. It's much more readable.

File: go/store/prolly/tree/json_location.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	pathBytes = pathBytes[1:]
+	for len(pathBytes) > 0 {
+		var lastIndex int
+		if pathBytes[0] == '[' {
+			lastIndex = bytes.IndexByte(pathBytes, ']')
+			arrayIndex, err := strconv.Atoi(string(pathBytes[1:lastIndex]))
+			if err != nil {
+				return path, err
+			}
+			path.appendArrayIndex(uint64(arrayIndex))
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: You aren't missing anything. This was a bug.
Entire implementation has been removed.

File: go/store/prolly/tree/json_location.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+		} else if pathBytes[0] == '.' {
+			pathBytes = pathBytes[1:]
+			lastIndex = min(bytes.IndexByte(pathBytes, '['),
+				bytes.IndexByte(pathBytes, '.'))
+			if lastIndex == -1 {
+				path.appendObjectKey(pathBytes)
+				return path, err
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: Implementation has been removed.

File: go/store/prolly/tree/json_location.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+func (p *jsonLocation) setScannerState(s scannerState) {
+	p.key[0] = byte(s)
+}
+
+func (p *jsonLocation) getScannerState() scannerState {
+	return scannerState(p.key[0])
+}
+
+func (p *jsonLocation) getPathElement(i int) (key []byte, isArray bool) {
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: I'm neutral, so I went ahead and made it a struct.

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f820432</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f820432</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/prolly/tree/node_cursor.go [hunk 10]
```diff
@@ -311,7 +314,11 @@ func recursiveFetchLeafNodeSpan(ctx context.Context, ns NodeStore, nodes []Node,
 }
 
 func currentCursorItems(cur *cursor) (key, value Item) {
-	key = cur.nd.keys.GetItem(cur.idx, cur.nd.msg)
+	if cur.nd.keys.IsEmpty() {
+		key = cur.parent.CurrentKey()
+	} else {
+		key = cur.nd.keys.GetItem(cur.idx, cur.nd.msg)
+	}
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/node_cursor.go:326]: Added comment.

File: go/store/prolly/tree/json_location.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+			path.appendArrayIndex(uint64(arrayIndex))
+		} else if pathBytes[0] == '.' {
+			pathBytes = pathBytes[1:]
+			lastIndex = min(bytes.IndexByte(pathBytes, '['),
+				bytes.IndexByte(pathBytes, '.'))
+			if lastIndex == -1 {
+				path.appendObjectKey(pathBytes)
+				return path, err
+			}
+			path.appendObjectKey(pathBytes[:lastIndex])
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: Implementation has been removed.

File: go/store/prolly/tree/json_location.go [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+// compareJsonLocations creates an ordering on locations by determining which one would come first in a normalized JSON
+// document where all keys are sorted lexographically.
+func compareJsonLocations(left, right jsonLocation) int {
+	minLength := min(left.size(), right.size())
+	for i := 0; i < minLength; i++ {
+		l, _ := left.getPathElement(i)
+		r, _ := right.getPathElement(i)
+		c := bytes.Compare(l, r)
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_location.go]: Not really, since it shouldn't ever be possible to compare one against the other; you can't have a value with both keys and indexes.

In practice, MySQL does have some corner cases around indexing into non-arrays, but I don't think here would ever be the place to detect those anyway.

File: go/store/prolly/tree/json_indexed_document.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+			return getInterfaceFromIndexedJsonMap(m)
+		}),
+	}
+}
+
+// Clone implements types.MutableJSON. Mutating an IndexedJsonDocument always returns a new IndexedJsonDocument without
+// modifying the state of the original. But creating a new instance allows callers to modify the value returned by ToInterface()
+// on the "mutable" copy without affecting the value cached on the original.
+func (i IndexedJsonDocument) Clone() types.MutableJSON {
+	// TODO: Add context parameter to MutableJSON.Clone()
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_indexed_document.go]: Yeah. I think I was punting on this until I have everything else working because I'd have to change a GMS API. I'll resolve this comment once I make that change.

File: go/store/prolly/tree/json_cursor.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+	result = append(result, jsonBuffer[startPos:j.jsonScanner.valueOffset]...)
+	return
+}
+
+func (j *JsonCursor) AdvanceToLocation(ctx context.Context, path jsonLocation) error {
+	// TODO: Intelligently determine whether the location exists in the current block, or whether we need to fetch a new block.
+	// Note for anyone debugging this: An infinite loop here means we're scanning the same block over and over again
+	// and not finding the requested location.
+
+	for compareJsonLocations(j.jsonScanner.currentPath, path) < 0 {
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_cursor.go]: Yeah I didn't like this either. It has a new implementation now that's better and should be easier to follow.

File: go/store/prolly/tree/json_chunker.go (hunk 4 above)
Comment (nicktobey) [reply go/store/prolly/tree/json_chunker.go]: You're right. Given that this function is only called in one place and isn't called anywhere else, I decided to just inline it and add comments.

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>443e348</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>443e348</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/prolly/tree/json_indexed_document.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_indexed_document.go]: I added a context parameter to clone, since it's significantly more tractable than adding a context to ToInterface. It's not ideal since while it works now, in the future ToInterface could be called with a different context. But it's good enough for now.

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1e72313</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1e72313</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8113baa</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8113baa</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>87ac817</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>87ac817</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>e4a6d6e</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>e4a6d6e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>
